examples/ensemble/plot_hgbt_regression.py

```diff
@@ -0,0 +1,397 @@
+# daily electricity transfer using the whole dataset. Then we visualize its
+# predictions depending on the `max_iter` parameter.
+
+from sklearn.ensemble import HistGradientBoostingRegressor
+
+max_iter_list = [10, 50]
+
+fig, ax = plt.subplots(figsize=(10, 5))
+average_week_demand = df.groupby(["day", "period"], observed=False)["transfer"].mean()
+average_week_demand.plot(color=colors[0], label="training data", linewidth=2, ax=ax)
+
+for idx, max_iter in enumerate(max_iter_list):
+    hgbt = HistGradientBoostingRegressor(max_iter=max_iter)
+    hgbt.fit(X, y)
+    y_pred = hgbt.predict(X)
+    prediction_df = df.copy()
+    prediction_df["y_pred"] = y_pred
+    average_pred = prediction_df.groupby(["day", "period"], observed=False)[
+        "y_pred"
+    ].mean()
+    average_pred.plot(
+        color=colors[idx + 1], label=f"max_iter={max_iter}", linewidth=2, ax=ax
+    )
+ax.set(
+    title="Average daily energy transfer during the week",
+    xticks=[(i + 0.2) * 48 for i in range(7)],
+    xticklabels=["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"],
+    xlabel="Time of the week",
+    ylabel="Normalized energy transfer",
+)
+_ = ax.legend()
+
+# %%
+# With just a few iterations, HGBT models can achieve convergence (see
+# :ref:`sphx_glr_auto_examples_ensemble_plot_forest_hist_grad_boosting_comparison.py`).
+#
+# Instead of relying solely on `max_iter` to determine when to stop, the HGBT
+# implementations in scikit-learn support early stopping. With it, the model
+# uses a fraction of the training data as a validation set
+# (`validation_fraction`) and stops training if the validation score does not
+# improve (or degrades) after `n_iter_no_change` iterations up to a certain
+# `tol`.
+#
+# Notice that there is a trade-off between `learning_rate` and `max_iter`:
+# Generally, smaller learning rates require more iterations to converge to the
+# minimum loss, while larger learning rates might converge faster but are at
+# risk of overfitting.
+#
+# Indeed, a good practice is to tune the learning rate along with any other
+# hyperparameters, fit the HBGT on the training set with a large enough value
+# for `max_iter` and determine the best `max_iter` via early stopping and some
+# explicit `validation_fraction`.
+
+common_params = {
+    "max_iter": 1_000,
+    "learning_rate": 0.3,
+    "validation_fraction": 0.2,
+    "random_state": 42,
+    "scoring": "neg_root_mean_squared_error",
+}
+
+hgbt = HistGradientBoostingRegressor(early_stopping=True, **common_params)
+hgbt.fit(X, y)
+plt.plot(-hgbt.validation_score_)
+plt.xlabel("number of iterations")
+plt.ylabel("root mean squared error")
+_ = plt.title(f"Loss of hgbt with early stopping (n_iter={hgbt.n_iter_})")
+
+# %%
+# We can then overwrite the value for `max_iter` to a razonable value and avoid
+# the extra computational cost of the inner validation. In this case, rounding
+# up the number of iterations to 600 may account for variability of the training
```

Comment:
That was my understanding of https://github.com/scikit-learn/scikit-learn/pull/26991#discussion_r1346051597. I may have got it wrong.

Reply:
Nop this is what @lorentzenchr mention. `max_iter` depends of the dataset and thus fixing it does not seem like a good practice to me if in the following the dataset is going to change (because you do cross-validation or just change the training set).

Is it really costly to have the validation of the early stopping always on?

Reply:
GBTs training is usually best done in the following order:
1) Find reasonable learning rate and max_iter, using early stopping (max iter just must be large enough such that early stopping triggers).
2) Optimize/Tune the rest of the hyperparameters (with fixed learning rate) with early stopping, also save the number of boosting iterations.
3) Retrain on whole training set with all parameters fixed, max_iter=number of boosting iterations (of best parameter set) and early stopping=False.

The reason is that the learning rate and max iter have a huge correlation.

Reply:
See also 4.4 in https://mayer79.github.io/statistical_computing_material/5_Trees.html#4_Gradient_Boosted_Trees